File: 2026-backendJS-week4/w4.py
# 購物車資料儲存
cart = []

# 運費規則設定
shipping_rules = {
    "free": {"min": 2000, "max": float('inf'), "fee": 0, "desc": "（免運）"},
    "standard": {"min": 1000, "max": 1999, "fee": 60, "desc": ""},
    "basic": {"min": 0, "max": 999, "fee": 80, "desc": ""}
}

def add_to_cart(original_price, sale_price, quantity):
    """第一階段：加入購物車並驗證"""
    is_valid = (
        original_price > 0 and 
        sale_price > 0 and 
        quantity > 0 and 
        isinstance(quantity, int) and 
        sale_price <= original_price
    )
    
    if not is_valid:
        print("輸入資料有誤，請重新輸入")
        return
    
    cart.append({
        "original_price": original_price,
        "sale_price": sale_price,
        "quantity": quantity
    })
    print(f"已加入購物車：商品原價 {original_price} 元，售價 {sale_price} 元，數量 {quantity} 件")

def calculate_total():
    """第二階段：計算總金額與運費"""
    if not cart:
        print("購物車是空的，請先加入商品")
        return

    total_original = sum(item["original_price"] * item["quantity"] for item in cart)
    total_sale = sum(item["sale_price"] * item["quantity"] for item in cart)
    savings = total_original - total_sale

    # 判定運費
    fee = 0
    desc = ""
    for rule in shipping_rules.values():
        if rule["min"] <= total_sale <= rule["max"]:
            fee = rule["fee"]
            desc = rule["desc"]
            break
            
    final_payable = total_sale + fee
    print(f"購物車原價總計 {total_original} 元，售價總計 {total_sale} 元，省下 {savings} 元，運費 {fee} 元{desc}，最終應付 {final_payable} 元")

def clear_cart():
    """第三階段：清空購物車"""
    cart.clear()
    print("購物車已清空")

# --- 測試流程 ---

# 第一階段測試
add_to_cart(1000, 1000, 2)
add_to_cart(800, 720, 3)
add_to_cart(2500, 2000, 1)
add_to_cart(0, 500, 2)
add_to_cart(1000, 800, 0)
add_to_cart(500, 600, 1)

# 第二階段測試
calculate_total()

# 第三階段測試
clear_cart()
add_to_cart(500, 450, 2)
add_to_cart(300, 280, 3)
calculate_total()
clear_cart()
calculate_total()

File: 2026-backendJS-week4/w4.js
let cart = []

// 1st step 
// 功能:加入購物車
function addToCart(originalPrice, salePrice, quantity){
  const isValid =
    originalPrice > 0 &&
    salePrice > 0 &&
    quantity > 0 &&
    salePrice <= originalPrice;

  if (!isValid) {
    console.log("輸入資料有誤，請重新輸入");
    return;
  }

  cart.push({originalPrice, salePrice, quantity});
  console.log(`已加入購物車:商品原價 ${originalPrice} 元，售價 ${salePrice} 元，數量 ${quantity}`);
}

// 2nd step
// 運費計算規則
const shippingRules = {
  free:{minAmount:2000, maxAmount:Infinity, fee:0, description:"免運費"},
  standard:{minAmount:1000, maxAmount:1999, fee:60, description:"折扣運費60"},
  basic:{minAmount:0, maxAmount:999, fee:80, description:"基本運費80"}
}

// 功能：運費計算
function calculateTotal(){
  // edge case
  if (cart.length === 0){
    console.log("購物車是空的，請先加入商品");
    return;
  }

  let totalOriginal = 0;
  let totalSale = 0;

  cart.forEach( item => {
    totalOriginal += item.originalPrice * item.quantity;
    totalSale += item.salePrice * item.quantity;
  });

  const savings = totalOriginal - totalSale;

  //運費判斷
  let shippingFee = 0;
  let shippingDesc = "";

  let rule = null;

  if (totalSale >= shippingRules.free.minAmount && totalSale <= shippingRules.free.maxAmount) {
    rule = shippingRules.free;
  } else if (totalSale >= shippingRules.standard.minAmount && totalSale <= shippingRules.standard.maxAmount) {
    rule = shippingRules.standard;
  } else if (totalSale >= shippingRules.basic.minAmount && totalSale <= shippingRules.basic.maxAmount) {
    rule = shippingRules.basic;
  }

  if (rule !== null) {
    shippingFee = rule.fee;
    
    if (rule.description !== "") {
      shippingDesc = `（${rule.description}）`;
    } else {
      shippingDesc = "";
    }
  }
  
  const finalAmount = totalSale + shippingFee;

  console.log(`購物車原價總計 ${totalOriginal} 元，售價總計 ${totalSale} 元，省下 ${savings} 元，運費 ${shippingFee} 元${shippingDesc}，最終應付 ${finalAmount} 元`);
}

// 3rd step 清空購物車
function clearCart() {
  cart = [];
  console.log("購物車已清空");
}


// --- 測試流程 ---

// 測試第一階段驗證與加入
addToCart(1000, 1000, 2);
addToCart(800, 720, 3);
addToCart(2500, 2000, 1);
addToCart(0, 500, 2);
addToCart(1000, 800, 0);
addToCart(500, 600, 1);

// 測試第二階段計算
calculateTotal();

// 測試第三階段完整流程
clearCart();
addToCart(500, 450, 2);
addToCart(300, 280, 3);
calculateTotal();
clearCart();
calculateTotal();
